tests/test_0163_negative-axis-wrap.py

```diff
@@ -0,0 +1,94 @@
+        assert awkward1.num(array, axis=3)
+    assert str(err.value) == "'axis' out of range for 'num'"
+
+    assert awkward1.to_list(awkward1.num(array, axis=-1)) == [[2, 2, 2, 2, 2],
+                                                              [2, 2, 2, 2, 2],
+                                                              [2, 2, 2, 2, 2]]
+    assert awkward1.to_list(awkward1.num(array, axis=-2)) == [5, 5, 5]
+    assert awkward1.num(array, axis=-3) == 3
+
+    with pytest.raises(ValueError) as err:
+        assert awkward1.num(array, axis=-4)
+    assert str(err.value) == "axis == -4 exceeds the depth == 3 of this array"
+
+def test_list_array():
+    array = awkward1.Array(numpy.arange(3*5*2).reshape(3, 5, 2).tolist())
+    assert awkward1.num(array, axis=0) == 3
+    assert awkward1.num(array, axis=1) == [5, 5, 5]
+    assert awkward1.num(array, axis=2) == [[2, 2, 2, 2, 2],
+                                           [2, 2, 2, 2, 2],
+                                           [2, 2, 2, 2, 2]]
+
+    with pytest.raises(ValueError) as err:
+        assert awkward1.num(array, axis=3)
+    assert str(err.value) == "'axis' out of range for 'num'"
+
+    assert awkward1.num(array, axis=-1) == [5, 5, 5]
+    assert awkward1.num(array, axis=-2) == [[2, 2, 2, 2, 2],
+                                           [2, 2, 2, 2, 2],
+                                           [2, 2, 2, 2, 2]]
+    assert awkward1.num(array, axis=-3) == 3
+    with pytest.raises(ValueError) as err:
+        assert awkward1.num(array, axis=-4)
+    assert str(err.value) == "axis == -4 exceeds the depth == 3 of this array"
+
+def test_record_array():
+    array = awkward1.Array([
+            {"x": [1], "y": [[], [1]]},
+            {"x": [1, 2], "y": [[], [1], [1, 2]]},
+            {"x": [1, 2, 3], "y": [[], [1], [1, 2], [1, 2, 3]]}])
+
+    assert awkward1.num(array, axis=0).tolist() == {'x': 3, 'y': 3}
+    assert awkward1.num(array, axis=1).tolist() == [{'x': 1, 'y': 2},
+                                                    {'x': 2, 'y': 3},
+                                                    {'x': 3, 'y': 4}]
+    with pytest.raises(ValueError) as err:
+        assert awkward1.num(array, axis=2)
+    assert str(err.value) == "'axis' out of range for 'num'"
+
+    assert awkward1.num(array, axis=-1).tolist() == [{'x': 1, 'y': [0, 1]},
+                                                     {'x': 2, 'y': [0, 1, 2]},
+                                                     {'x': 3, 'y': [0, 1, 2, 3]}]
+
+def test_record_array_axis_out_of_range():
+    array = awkward1.Array([
+            {"x": [1], "y": [[], [1]]},
+            {"x": [1, 2], "y": [[], [1], [1, 2]]},
+            {"x": [1, 2, 3], "y": [[], [1], [1, 2], [1, 2, 3]]}])
+
+    # FIXME: this should throw
+    a = awkward1.num(array, axis=-2)
+    # <RecordArray>
+    # <field index="0" key="x">
+    #     <NumpyArray format="l" shape="" data="0x 03000000 00000000" at="0x7fd818c06c60"/>
```

Comment:
@jpivarski - it looks like `to_list` cannot handle a `NumpyArray` with no shape.

Reply:
A NumpyArray with `len(shape) == 0` should never get to the Python level. `Content` subclasses sent to Python should pass through this `box` function ([here](https://github.com/scikit-hep/awkward-1.0/blob/5f1a068fec0d00936bb353318911855788a655ec/src/python/content.cpp#L14-L130)) that turns a scalar NumpyArray (one with `len(shape) == 0`) into a Python number. The same is true of `None` (the subclass of `Content` that should turn into a Python `None`). If this is what you're seeing, the error is upstream.

All of the functions that return `Content` subclasses in src/python/content.cpp are _supposed to_ pass their output through `box`, but there isn't anything in place to ensure this is the case.

Reply:
Ah! Thanks, @jpivarski! More debugging then :-)